npm/pkg/controlplane/translation/parseSelector.go

```diff
@@ -238,3 +238,164 @@ func parseSelector(selector *metav1.LabelSelector) (labels []string, vals map[st
 
 	return labels, vals
 }
+
+// labelSelector has parsed matchLabels and MatchExpressions information.
+type labelSelector struct {
+	// include is a flag to indicate whether Op exists or not.
+	include bool
+	settype ipsets.SetType
+	// setName is among
+	// 1. matchKey + ":" + matchVal (can be empty string) case
+	// 2. "matchKey" case
+	// or 3. "matchKey + : + multiple matchVals" case.
+	setName string
+	// members slice exists only if setType is only NestedLabelOfPod.
+	members []string
+}
+
+// parsedSelectors maintains slice of unique labelSelector.
+type parsedSelectors struct {
+	labelSelectors []labelSelector
+	// Use set data structure to avoid the duplicate setName among matchLabels and MatchExpression.
+	// The key of labelSet includes "!" if operator is "OpNOtIn" or "OpDoesNotExist"
+	// to make difference when it has the same key (and value), but different operator
+	// while this is weird since it is not always matched, but K8s accepts this spec.
+	labelSet map[string]struct{}
+}
+
+func newParsedSelectors() parsedSelectors {
+	return parsedSelectors{
+		labelSelectors: []labelSelector{},
+		labelSet:       map[string]struct{}{},
+	}
+}
+
+// addSelector only adds non-duplicated labelSelector.
+// Only nested labels from podSelector has members fields.
+func (ps *parsedSelectors) addSelector(include bool, setType ipsets.SetType, setName string, members ...string) {
+	setNameWithOp := setName
+	if !include {
+		setNameWithOp = "!" + setName
+	}
+
+	// in case setNameWithOp exists in a set, do not need to add it.
+	if _, exist := ps.labelSet[setNameWithOp]; exist {
+		return
+	}
+
+	ls := labelSelector{
+		include: include,
+		settype: setType,
+		setName: setName,
+		members: members,
+	}
+
+	ps.labelSelectors = append(ps.labelSelectors, ls)
+	ps.labelSet[setNameWithOp] = struct{}{}
+}
+
+// parseNSSelector parses namespaceSelector and returns slice of labelSelector object
+// which includes operator, setType, ipset name and always nil members slice.
+// Member slices is always nil since parseNSSelector function is called
+// after FlattenNameSpaceSelector function is called, which guarantees
+// there is no matchExpression with multiple values.
+// TODO: good to remove this dependency later if possible.
+func parseNSSelector(selector *metav1.LabelSelector) []labelSelector {
+	// TODO(jungukcho): This will not happen
+	if selector == nil {
+		return []labelSelector{}
+	}
+
+	parsedSelectors := newParsedSelectors()
+
+	// #1. All namespaces case
+	if len(selector.MatchLabels) == 0 && len(selector.MatchExpressions) == 0 {
+		parsedSelectors.addSelector(true, ipsets.KeyLabelOfNamespace, util.KubeAllNamespacesFlag)
```

Comment:
would we want to add the all-namespaces selector for all namespaces? Why just conditionally? Sorry that I didn't question in your first PR.

If we want to for all, then we can remove the return statement for this situation since the loops will do nothing when lengths are 0.

Reply:
Sorry. I am not clear about question. Can you elaborate what you mean?
Code is for the case when  
```shell
 namespaceSelector: {}
```

Reply:
nevermind, I was confusing this with ipsets. Makes sense how we select all namespaces only if there are no matchLabels